### ISSUE TITLE
Option to select decisions as branches for coverage reports

### ISSUE THREAD

**Issue by chandryan** [gcovr/gcovr#1042]
# Is your feature request related to a problem?

We use gcovr to get the branch coverage of new code within PRs. We're often struggling with False Negatives and user code using some libraries like Boost or gRPC seems to be especially affected by this. The coverage metrics from `--decisions` look like an attractive alternative, but we can't make use of that because they are not considered as input for the coverage report (I'm using `--sonarqube`).

# Describe the solution you'd like

I would like to have an option to use decision coverage as the branch coverage for the XML report that is output with the `--sonarqube` option. I'm not sure about what exact content the other types of coverage reports supported by gcovr contain, I suppose they also don't use the decisions and one option could be generically used for this?

My solution proposal is to introduce an additional option like the following:

```
    --decisions-as-branches
Use the decision coverage as branch coverage for coverage reports, which cannot show both.
Implies --decisions. Used by x, y, and SonarQube reports.
```

# Describe alternatives you've considered

I thought about some variations of the option's impact, though the proposed solution is my favorite one.

*Alternative 1: An option only for the SonarQube report*

I suppose the same principle can be applied to other reports as well.

*Alternative 2: Use decisions as branches for all reports*

For reports like the HTML reports this is not necessary, as they can show both metrics. It would also be unnecessarily confusing to actually show the decisions column as a branches column. However, it might be worth a thought colouring the lines in HTML reports according to the decisions instead of the branches.

# Additional context

Happy to contribute and create a PR in case you agree with the idea 😄

**Comment by Spacetown** [comment]
The option could be reused but I'm not sure if I like the idea of changing the content of a report file by adding the "wrong" data to the elements. 
Also I wouldn't change the coloring of the html report.

@latk what do you think about this idea?

**Comment by Spacetown** [comment]
> However, I'm really not a fan of the `--decisions` feature. Unless GCC 14+ is used (where we can take condition-coverage data directly from the gcov JSON), the decision analyzer is super fragile, makes unreasonably strong assumptions about the formatting of the source code, and only works for .gcov file inputs. 

I suggest to add a warning if this feature is used without having the JSON files of gcc >= 14.

> I wonder how the claim "[`gcovr --decisions`] provides a reliable tool for (i.e. MISRA-compliant) code in security-relevant situations" made it into our docs, because that is blatantly false (except when using GCC 14+). Nothing about this feature is robust or reliable, and it cannot be relied upon for security-sensitive purposes or certifications (except maybe when using GCC 14+).

I suggest to remove this block from the FAQ.

**Comment by latk** [comment]
Hi @chandryan. It seems you're right, and I have misread the source code. The `DecisionParser` is used under all circumstances whenever `--decisions` is requested.

There was some wishful thinking on my part that we could just use the `conditions` data and relabel it as `decisions` and call it a day, which would be close enough to the current behavior in all scenarios where the `DecisionParser` actually works. Notably, `DecisionParser` is just a very basic filter to ignore branch coverage data that doesn't clearly relate to decisions (outcomes) of control flow constructs. In pseudocode, it more or less works like:

```python
if source_code.strip().replace(" ", "").startwith("if("):
   if len(cov.branches) == 2:
     cov.decisions = (cov.branches[0], cov.branches[1])
   else:
     cov.decisions = "uncheckable"
```

(actual code for single-line control flow [here](https://github.com/gcovr/gcovr/blob/545f18745134842a09e88cfbd3f579ae4ce4f257/src/gcovr/decision_analysis.py#L203-L220))

In all interesting cases where there would be a difference between decision coverage and condition coverage, `gcovr --decisions` *does not work*. The parser could be made *slightly* more robust by using condition-data instead of branch-data whenever available, but the fundamental issues would remain. Instead of checking whether the control flow line has two branches, it could check whether the line contains a single condition. You correctly point out that GCC produces info about condition coverage, and that there's no clear way to convert that into meaningful decision coverage (at least not without parsing and evaluating the boolean expressions).

On second thought, a more useful feature than "decisions-as-branches" could be "conditions-as-branches" (GCC 14+ only), which should be fairly accurate and intuitive.

**Comment by chandryan** [comment]
I played around with GCC14 and created some example code and coverage reports that I pushed [here](https://github.com/chandryan/coverage_playground/commit/4fd77ef66f807220e9125b80a89a932b61b48a6c) in case you're interested in also taking a look :-)

For relating the next points - the example code had 25 lines with 4 decisions (2 if-branches). All lines and 3/4 decisions were covered. I had the following discoveries (referring to gcc14 only, as gcc13 had a slightly different branch coverage output):

**Branch coverage**
The report showed 26/36=72% branch coverage. That metric seems quite meaningless with respect to what a user would expect to be able to get from branch coverage (which I assume to be the decision coverage). All function calls seem to add at least one branch, distorting the metric so much that the absolute numbers don't make sense at all. It is surprisingly close to the real 75% percent, though looking at the locations where uncovered branches occur I suspect this to be a coincidence.

**Condition coverage**
There was 8/14=57% condition coverage. There were two reasons why the result deviates that much from the actual expected coverage of 3/4=75%:
- it is indeed MC/DC coverage, which is great if you really need it but potentially painful if you're just interested in decision coverage, due to again getting a lower coverage metric than what you're actually interested in
- One variable declaration `std::vector<int> v = {1, 2, 3, 4, 5};`, that does not contain conditions, for some unknown reason resulted in a 4/8 condition coverage output.

From a perspective of wanting the best approximate for a decision coverage metric I see little benefit of using the condition coverage over the current `--decisions` way: Re-interpreting the MC/DC metric means you have to parse the source code with regexes again, while the other artifact also implies that you cannot trust the condition metric of GCC to reliably relate to source code instead of some assembly code. Which again means you need to fall back to parsing source code to exclude such situations.

**Decision coverage**
The metric is correctly showing a 3/4=75% coverage, though admittedly the file was easy to parse.


Despite the "ugly" method of the `DecisionParser` using reg-exes on text instead of using information coming from the compiler directly, my feeling goes more and more into the direction that this might not be such a bad idea after all. At least if the compiler doesn't offer you a better way, and seeing that the amount of patterns required to identify control structures is limited. There's even a famous tool [cpplint](https://github.com/cpplint/cpplint) linting C++ code only by parsing text, and it does a fairly decent job.


After having gone through these experiments, from my point of view a `--decisions-as-branches` still seems useful 😄
What do you guys think?

**Comment by chandryan** [comment]
Ohh and I just noticed one more problem with the branch coverage metric:

So far I've thought that, while being inaccurate, it would at least show you a result which is not higher than the actual coverage. But looking at the gcc13 branch coverage result it shows 23/29=79%, which is more than the actual 75% :-[

So while it's true that you don't get False Positives for branch coverage on a line, due to the distorted metric resulting from "branch" coverage from function calls etc. the percentages can still show misleading results.

**Comment by Spacetown** [comment]
The condition coverage from GCC is MC/DC coverage. In your example the `std::vector<int> v = {1, 2, 3, 4, 5};` seems to use a condition (inlined constructor?). There are also conditions inserted for function static variables, if you really only want to see the coverage for the user code, the only solution is to parse the code.

With GCC-14 we now also have the target branches in the data model, if the target branch of the different condition branches are analysed it should be possible to get the DC without parsing the source.

**Comment by chandryan** [comment]
> The condition coverage from GCC is MC/DC coverage. In your example the `std::vector<int> v = {1, 2, 3, 4, 5};` seems to use a condition (inlined constructor?). There are also conditions inserted for function static variables, if you really only want to see the coverage for the user code, the only solution is to parse the code.
> 
> With GCC-14 we now also have the target branches in the data model, if the target branch of the different condition branches are analysed it should be possible to get the DC without parsing the source.

Thanks for the additional input!
I spent more time today to look further into the options given by GCC-14. As much as I would have loved to have a perfect solution with 100% accurate metrics, I'm ready to give up now :(
I couldn't understand how to identify target branches from a potential branch/condition metric correlation in the gcov output to tackle the MC/DC challenge, but as you stated, the main issue remains: I want to only see the coverage of the user code and the only solution is to parse it. In my context using GCC-14 would probably also not have been an option for a long time.

**To summarize the situation:**

I have a use case related to passing Quality Gates exactly as you mentioned, and in my context an "optimistic" metric using decisions as branches with False Positives would indeed be acceptable - and the preferred one over the headaches we've been experiencing for some PRs that would not pass the Quality Gate, despite proper coverage.

Implementation-wise it seems rather easy to choose one or the other metric as input for the "branch" coverage in the XML reports. I would be happy to contribute by providing such a switch, though I understand if you prefer not to have this option available.

**Comment by Spacetown** [comment]
@chandryan Are you still working on it?
